ledger/participant-integration-api/src/test/suite/scala/platform/store/cache/EventsBufferSpec.scala

```diff
@@ -0,0 +1,156 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.platform.store.cache
+
+import com.codahale.metrics.MetricRegistry
+import com.daml.metrics.Metrics
+import com.daml.platform.store.cache.EventsBuffer.{RequestOffBufferBounds, UnorderedException}
+import org.scalacheck.Gen
+import org.scalatest.compatible.Assertion
+import org.scalatest.matchers.should.Matchers
+import org.scalatest.wordspec.AnyWordSpec
+import org.scalatest.{Failed, Succeeded}
+import org.scalatestplus.scalacheck.ScalaCheckDrivenPropertyChecks
+
+import scala.collection.immutable
+
+class EventsBufferSpec extends AnyWordSpec with Matchers with ScalaCheckDrivenPropertyChecks {
+  private val BufferElements = Vector(2, 3, 5, 8, 13).map(idx => idx -> idx * 2)
+  private val LastOffset = BufferElements.last._1
+
+  "push" when {
+    "max buffer size reached" should {
+      "drop oldest" in withBuffer(3) { buffer =>
+        buffer.slice(0, LastOffset) shouldBe BufferElements.drop(2)
+        buffer.push(21, 42)
+        buffer.slice(0, 21) shouldBe BufferElements.drop(3) :+ 21 -> 42
+      }
+    }
+
+    "element with smaller offset added" should {
+      "throw" in withBuffer(3) { buffer =>
+        intercept[UnorderedException[Int]] {
+          buffer.push(1, 2)
+        }.getMessage shouldBe s"Elements appended to the buffer should have strictly increasing offsets: 13 vs 1"
+      }
+    }
+
+    "element with equal offset added" should {
+      "throw" in withBuffer(3) { buffer =>
+        intercept[UnorderedException[Int]] {
+          buffer.push(13, 2)
+        }.getMessage shouldBe s"Elements appended to the buffer should have strictly increasing offsets: 13 vs 13"
+      }
+    }
+
+    "range end with equal offset added" should {
+      "accept it" in withBuffer(3) { buffer =>
+        buffer.push(13, Int.MaxValue)
+        buffer.slice(0, 13) shouldBe BufferElements.drop(2)
+      }
+    }
+
+    "range end with greater offset added" should {
+      "not allow new element with lower offset" in withBuffer(3) { buffer =>
+        buffer.push(15, Int.MaxValue)
+        buffer.slice(0, 13) shouldBe BufferElements.drop(2)
+        intercept[UnorderedException[Int]] {
+          buffer.push(14, 28)
+        }.getMessage shouldBe s"Elements appended to the buffer should have strictly increasing offsets: 15 vs 14"
+      }
+    }
+  }
+
+  "getEvents" when {
+    "called with valid range" should {
+      val bufferElements = increasinglySparse(10).map(offset => offset -> offset * 2)
+
+      "return the correct series" in withBuffer(10, bufferElements) { buffer =>
+        val rangesGen = for {
+          startExclusive <- Gen.choose(0, bufferElements.last._1)
+          endInclusive <- Gen.choose(startExclusive, bufferElements.last._1)
+        } yield (startExclusive, endInclusive)
```

Comment:
I am not sure how exhaustive those generators are, I would be very interested in a pinpointed test that explicitly hits the buffer at its extremes i.e. returns the full buffer

Reply:
Fine with that.

Reply:
Done.